### ISSUE TITLE
Memory leak causing server crash when parsing large file [optimisation]

### ISSUE THREAD

**Issue by nevadascout** [HvyIndustries/crane#157]
In the Yii 1.1.17 framework there is an included vendor file from the `htmlpurifier` library which is over 21,800 lines long.

Watching the memory usage in the task manager, the vs code server process spawns, quickly climbs to nearly 1GB of memory usage then crashes out.

Not sure what the solution here is. Do we simply not parse files longer than, say, 10,000 lines?

Longer term we should optimise the language server to cache these larger files to disk during the parsing process to avoid memory overflows.

_Framework: **Yii 1.1.17**_

**Comment by ichiriac** [comment]
Hi,

I'm investigating on this, it seems that's a nasty memory problem on lexer (I'm thinking at his revert backtrace memory) 

For example, just reading the same 100 first tokens 10k times, and using the same storing structure, i get 8 Mb of memory usage. Doing the same and reading each token 10k times, I get a memory problem at 560 Mb.

You can't really limit in terms of number of lines, it's not the real limitation, but instead in terms of tokens, an easy way would be to limit the file size, lets say 10 Mo max by file (about 100 Kloc) - considering it's like any average file with some code, some comments, and some texts.

**Comment by ichiriac** [comment]
by commenting this line, it will fix the memory link :
https://github.com/glayzzle/php-parser/blob/master/src/lexer.js#L259

This line causes a memory leak only from this lexer call :
https://github.com/glayzzle/php-parser/blob/master/src/lexer/php_main.l#L181

I've verified, the char is inserted once, and the _input string keeps decreasing. It's like NodeJS allocates a new string as a result, and does not collect the old un-referenced string from memory. 

I have no clue, why a simple string manipulation would make a memroy leak.

**Comment by nevadascout** [comment]
>It's like NodeJS allocates a new string as a result, and does not collect the old un-referenced string from memory.

Can you manually unset it or remove it using `delete` instead of relying on the garbage collection?

**Comment by ichiriac** [comment]
Hi @nevadascout,

I've tried all tweaks, and it does not works. I've also tried to isolate the behaviour to reproduce the memory leak but I've not succeed.

I've also undone all my customisations to keep the lexer as originally generated and to check if I've introduced these behaviours, and the original version also contains the memory leak.

Anyway, the lexer actually has poor performance, so I plan to rewrite it entirely, I should gain at least 2x speed and a much lower memory consumption (even if there wasn't a memory leak).

The only way to force parsing of a large file like tcpdf.php (about 900k) is to add `--max_old_space_size=6000` tag on node

**Comment by stevenwood2014** [comment]
Just an FYI, I too am experiencing this issue when indexing my PHP workspace.
It was crashing due to one particular vendor file that was actually around 936kb in size, but it sounds like perhaps due to its complexity it is causing the parser to crash.

The offending file (in case it helps track down the underlying issue with the indexer/parser) is:

 @package    PHPExcel
 @copyright  Copyright (c) 2006 - 2011 PHPExcel (http://www.codeplex.com/PHPExcel)
 @version    1.7.6, 2011-02-27
 @file   PHPExcel/Shared/PDF/tcpdf.php

**Comment by nevadascout** [comment]
Thanks @stevenwood2014! This TCPDF file is actually the same one that is crashing for me.

I'm not 100% sure how best to solve this issue _(short of fixing the memory leak)_... Anyone have any suggestions?

**Comment by ichiriac** [comment]
I've partially rewrite the tokenizer, here some early results :

## With a 92kb file :

### Old version 
Duration :  206 ms
Memory :  4465 kb

### New Version
Duration :  49 ms
Memory :  1592 kb

## With a 664kb file :

### Old version 
Duration :  735ms
Memory :  16352 kb

### New Version
Duration :  335 ms
Memory :  11003 kb

## With a 1280kb file :

### Old version 
Duration :  3426 ms
Memory :  32508 kb

### New Version
Duration :  855 ms
Memory :  18932 kb

---

## The new version should be 
- About 3x more rapid
- Use 2x less memory


The code used for tests is a copy/paste of this :

```php
Hello world <?= $foo; ?> and <?= $$foo; ?> or <?= ${$foo}; ?>
<?php

# Old fashion Style Comment

  /**
   * Multi Line Comment
   */
  echo '
    TEST \'AZERTY\' \r\n\t
  ';      ?>
```

A stable release should come in few weeks

**Comment by ichiriac** [comment]
Just released php-parser@0.1.0, this bug is now fixed (more details here https://github.com/glayzzle/php-parser/issues/19)

**Comment by ichiriac** [comment]
Now the parser/lexer in no more static / only instances - no more memory leaks. Tested over 60k files without any extra memory consumption.

Also tested and passes with a single parser instance that parses many files, without any memory leak.

Be aware the garbage collector does not automatically collects at each cycle, so the memory can increase during massive parsing by some 100/200Mo and release the memory just few seconds after.

The parser is synchronous as it's processor intensive, so if the files reading loop is also synchronous it loop block nodejs loop blocking also garbage collector cycles.

A good practice would be :

* Looping on files in async mode
* Create a new instance of the parser for each file parsing
* In order to process in parallel, use a fork/thread model (as the parsing is done in a single thread/core)

---

I close this issue as it's fixed, if you find any memory problems with the next release you  can open a new issue.

**Comment by nevadascout** [comment]
Released in v0.3.0
